Introduce session cache

zhangrui1203 opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.io to discuss further. -->

## Motivation

SGLang provides session-level KV cache management, enabling user-defined business logic to explicitly control advanced features such as the lifecycle of the KV cache, hierarchical caching, and replica count management. In large-scale LLM service scenarios within data centers, SGLang serves as a high-performance data processing node, better supporting business logic and achieving superior results in performance, cost efficiency, and user experience.

## Modifications

- implement session cache framework
- improve hicache local file to support session cache
- support GD2FS as session cache storage backend
- improve performance for get_dummy_flat_data_page
- bug fix for cache_finished_req

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

zhangrui1203 committed:
Batch allocate dummy flat data pages to speed up large KV cache initialization

Previously, `get_dummy_flat_data_page` allocated a single page at a
time, causing significant overhead when initializing a large KV cache.
This change introduces batch allocation to reduce initialization time.

Signed-off-by: zhangrui <zhangrui@tensorfer.com>
Signed-off-by: zhenwei pi <pizhenwei@tensorfer.com>

zhangrui1203 committed:
Fix the size of the KV cache written back

Signed-off-by: zhangrui <zhangrui@tensorfer.com>

zhangrui1203 committed:
Introduce session cache

In previous versions, sglang instances managed the KV cache directly
and achieved notable improvements by integrating HiCache storage for
limited scalability.

We now introduce a new session-based cache management mechanism that
enables upper-layer business logic to fine-grain control KV cache
behavior. For example, business logic can dynamically adjust the
expiration time of a user’s session KV cache based on their priority,
thereby delivering differentiated user experiences.

Co-authored-by: Zhiqiang Xie <xiezhq@stanford.edu>
Co-authored-by: wangyu <wangyu.steph@bytedance.com>
Signed-off-by: zhangrui <zhangrui@tensorfer.com>
Signed-off-by: zhenwei pi <pizhenwei@tensorfer.com>

zhangrui1203 committed:
Introduce GD2FS as HiCache Storage backend

Only the save/load interface is supported for now,
used for the backend storage of session-cache.

Signed-off-by: zhangrui <zhangrui@tensorfer.com>
Signed-off-by: zhenwei pi <pizhenwei@tensorfer.com>

zhangrui1203 committed:
Optimize the get_flat_data in the memory pool

Signed-off-by: zhangrui <zhangrui@tensorfer.com>
Signed-off-by: zhenwei pi <pizhenwei@tensorfer.com>

zhangrui1203 committed:
Restore the number of torch threads to speed up the tensor allocation on the host

Signed-off-by: zhangrui <zhangrui@tensorfer.com>
Signed-off-by: zhenwei pi <pizhenwei@tensorfer.com>

xiezhq-hermann commented on python/sglang/srt/managers/io_struct.py:
```diff
@@ -143,6 +144,21 @@ class SessionParams:
+    def __post_init__(self):
+        old = SessionCache(self.old_kv_cache)
+        new = SessionCache(self.new_kv_cache)
+        if len(old) > 0 and len(new) > 0 and old.token_end != new.token_start:
```
this is a very strong assumption of execution, do you have following plans on fault tolerance?

zhangrui1203 replied:
If users intend to utilize the session cache (by specifying session_id along with stored_kv_cache and fresh_kv_cache), they must strictly adhere to the following requirements:

1. The internal segments of both stored_kv_cache and fresh_kv_cache must be contiguous in terms of sequence positions.
2. The fresh_kv_cache must be directly appendable to the tail of the stored_kv_cache, maintaining sequential continuity.

If these conditions are not satisfied, SGLang must return an error. This is because the session cache relies on metadata that is managed and persisted by the control plane (i.e., the user). As a data-plane component, SGLang cannot interpret or recover from malformed or inconsistent control-plane metadata.

xiezhq-hermann commented on python/sglang/srt/managers/schedule_batch.py:
```diff
@@ -1003,7 +1009,7 @@ def _check_str_based_finish(self):
                 for stop_regex_str in self.sampling_params.stop_regex_strs:
                     if re.search(stop_regex_str, tail_str):
-                        self.finished_reason = FINISHED_MATCHED_REGEX(
+                        self.finished_reason = FINISH_MATCHED_REGEX(
```
let's have all sorts of style fix in a separate PR

zhangrui1203 replied:
OK

xiezhq-hermann commented on python/sglang/srt/managers/scheduler.py:
```diff
@@ -1968,6 +1999,13 @@ def get_new_batch_prefill(self) -> Optional[ScheduleBatch]:
+                if session_id is not None:
+                    done = self.tree_cache.check_session_prefetch_progress(session_id)
+                    if not done:
+                        continue
```
so here is a blocking call for prefetching for the session cache, will there be more policies for it?

zhangrui1203 replied:
Currently, a generate request will block indefinitely until the session cache prefetch operation completes. In the future, timeout support will be added to prevent the generate request from hanging due to excessively long prefetch operations; however, this capability depends on the implementation of the backend storage system.

xiezhq-hermann commented on python/sglang/srt/mem_cache/session_cache.py:
```diff
@@ -0,0 +1,615 @@
+
+            storage.load(filepath, seg.kv_start, flat_data)
+
+            mem_pool_host.set_from_flat_data(mem_indices, flat_data)
```
will there be a zero-copy mechanism for loading data from the storage?

xiezhq-hermann commented on python/sglang/srt/mem_cache/session_cache.py:
```diff
@@ -0,0 +1,615 @@
+                seg.kv_uri, storage_config, mem_pool_host
+            )
+
+            storage.save(filepath, seg.kv_start, seg_data)
```
same comments for zero-copy

xiezhq-hermann commented on python/sglang/srt/mem_cache/radix_cache.py:
```diff
@@ -447,7 +453,7 @@ def cache_finished_req(self, req: Req, is_insert: bool = True):
 
         # Maybe convert to bigram keys for EAGLE
-        keys = convert_to_bigram_key(req.fill_ids) if self.is_eagle else req.fill_ids
+        keys = convert_to_bigram_key(token_ids) if self.is_eagle else token_ids
```
not sure about this one, if it's more of a style change please make it in a separate PR, otherwise will need @ispobock's opinion on it

zhangrui1203 replied:
OK

xiezhq-hermann commented:
Thank you @zhangrui1203 @pizhenwei for the PR, overall a very neat PR and improved a lot since my initial prototype: https://github.com/sgl-project/sglang/tree/xiezhq-session. I left some comments and have a few addition questions:
1. I do think the session cache abstraction can be compatible with other storage backends with only minor interface changes, e.g., providing a file interface to load and save session specific KV caches. It's ok for now to only support your backend but please help think about it in future refactoring.
2. Zero-copy from storage to memory is still missing, leaving fairly substantial performance on the table.
3. Can you prepare a test case just like other storage backends do, that way we have a correctness guarantee. It would be better to have a simple performance benchmark too, we can adapt the existing multi-turn benchmark script for it a bit. A validation on the session abstraction achieving close or even better performance over existing pages would be great.

zhangrui1203 commented on python/sglang/srt/mem_cache/memory_pool.py:
```diff
@@ -809,6 +816,31 @@ def get_cpu_copy(self, indices):
+            pin_memory=True,
+        )
+
+        torch.cuda.synchronize()
```
OK

zhangrui1203 commented:
Hi, @xiezhq-hermann 
The main changes in the new version are as follows:

- rename `old_kv_cache`/`new_kv_cache` to `stored_kv_cache`/`fresh_kv_cache`
- remove style fix
- use `torch.cuda.stream()` instead of `torch.cuda.synchronize()` in `get_flat_data`
- rebase to the latest commit

The test case is currently being added.

zhangrui1203 committed:
Add test case for session cache

Signed-off-by: zhangrui <zhangrui@tensorfer.com>

zhangrui1203 committed:
Add benchmark for session cache

Benchmark Environmant: NVIDIA A800-SXM4-40GB
> without session cache
 $ ./bench_session_cache.sh \
    -p 30000 \
    -m /models/Qwen3-8B/ \
    -d /dataset/ShareGPT_V3_unfiltered_cleaned_split.json \
    -b \
    --disable-session-cache

Performance metrics summary:
  Total requests: 2048 at 16 requests per second
  Average Prompt Length: 8854.15 tokens
  Average Output Length: 128.00 tokens
  Average TTFT: 6.33
  P90 TTFT: 11.45
  Median TTFT: 5.52
  Average latency: 13.91
  P90 latency: 21.62
  Median latency: 11.67
  Input token throughput: 9939.50 tokens per second
  Output token throughput: 143.69 tokens per second
  Request Throughput: 1.12 requests per second
  Cache Hit Rate: 0.103550

> with session cache
 $ ./bench_session_cache.sh \
    -p 30000 \
    -m /models/Qwen3-8B/ \
    -d /dataset/ShareGPT_V3_unfiltered_cleaned_split.json \
    -b

Performance metrics summary:
  Total requests: 2048 at 16 requests per second
  Average Prompt Length: 8854.14 tokens
  Average Output Length: 128.00 tokens
  Average TTFT: 4.45                                  -> reduce ~30%
  P90 TTFT: 7.47                                      -> reduce ~35%
  Median TTFT: 3.72                                   -> reduce ~32%
  Average latency: 7.86                               -> reduce ~43%
  P90 latency: 11.41                                  -> reduce ~47%
  Median latency: 6.78                                -> reduce ~42%
  Input token throughput: 17520.13 tokens per second  -> improve ~76%
  Output token throughput: 253.29 tokens per second   -> improve ~77%
  Request Throughput: 1.98 requests per second        -> improve ~77%
  Cache Hit Rate: 0.870952                            -> improve ~741%

Signed-off-by: zhangrui <zhangrui@tensorfer.com>
Signed-off-by: zhenwei pi <pizhenwei@tensorfer.com>

zhangrui1203 commented:
@xiezhq-hermann 
One more thing: While adding accuracy tests, we noticed they depend on @sgl.function, which would require API changes. Are there simpler or more standard ways to test this?

alphabetc1 commented on benchmark/hicache/bench_multiturn.py:
```diff
@@ -356,6 +376,58 @@ def __init__(self, args):
+
+        # Create session cache files
+        hicache_file_path = os.getenv(
+            "SGLANG_HICACHE_FILE_BACKEND_STORAGE_DIR", "/tmp/hicache"
```
nit: please put all environment variables in `environ.py` and use `envs.VAR_NAME.get()`

zhangrui1203 commented:
Hi, @xiezhq-hermann 
All CI runs have completed. The failures appear unrelated to this PR, and some may pass with a retry.

xiezhq-hermann commented:
@zhangrui1203 thank you for all your effort. How has the accuracy testing been going? My understanding is that as long as we’re using the same session, the session cache should be in effect.

Regarding the merge, I’d suggest splitting it into multiple independent PRs. That should make the code review easier and keep each commit atomic.
